### PR TITLE
[pvr] cleanups in PVR manager

### PR DESCRIPTION
**Pull request by xhaggi**
Rewrite of announce logic used by PVR manager + some small refactorings.

@ksooo mind taking a look.

### CODE REVIEW

**Comment by ksooo** [comment]
Well, this is not the right approach. @FernetMenta might explain this best, because he introduced the direct pvr manager calls in powermanager for a reason.

**Comment by xhaggi** [comment]
Okay please let me know why, i didn't noticed it or point me to the discussion. IMO it makes sense to decouple the PVR manager from the PowerManager because the announce is trigger right next to that lines.

**Comment by ksooo** [comment]
Ask fernet, please. I was not directly involved, and only remember that there was some discussion somewhere leading to a pvr api change and that fernet was involved. Sorry, seems I'm getting old. ;-)

**Comment by xhaggi** [comment]
The discussion is in #9662 which introduce the sleep/wakeup pvr client stuff.

**Comment by xhaggi** [comment]
The last commit will drop the announce logic for OnQuit, OnSleep, OnWake and use the methods `OnWake()`, `OnSleep()` and `Deinit()` instead. The methods are called directly like before, but I dropped the `SetWakeCommand()` call from the `PowerManager`. If good I'll squash the commit.

**Comment by ksooo** [comment]
Yeah, this reads good. Will take a look at the code for the usual nitpicking later, when I'm back home.

File: xbmc/pvr/PVRManager.cpp [hunk 1]
```diff
@@ -594,12 +573,31 @@ bool CPVRManager::SetWakeupCommand(void)
 
 void CPVRManager::OnSleep()
 {
+  SetWakeupCommand();
+  SaveLastPlayedChannel();
```
Comment (ksooo) [inline xbmc/pvr/PVRManager.cpp]: Just cosmetics, but in CPVRManager::Deinit() you are calling these two methods in the the opposite order.

File: xbmc/pvr/PVRManager.h [hunk 2]
```diff
@@ -561,6 +556,16 @@ namespace PVR
 
   private:
     /*!
+     * @brief Save the current playing channel as last played channel
```
Comment (ksooo) [inline xbmc/pvr/PVRManager.h]: correct English is 'currently' here, imo.

File: xbmc/pvr/PVRManager.cpp [hunk 3]
```diff
@@ -594,12 +573,31 @@ bool CPVRManager::SetWakeupCommand(void)
...
+
   g_PVRClients->OnSystemSleep();
 }
 
 void CPVRManager::OnWake()
 {
   g_PVRClients->OnSystemWake();
+
+  if (IsStarted())
```
Comment (ksooo) [inline xbmc/pvr/PVRManager.cpp]: I just spotted, that CPVRManager::Announce always() does an IsStarted check before doing anything.

Okay, but then you should add the IsStarted check also to CPVRManager::OnSleep() and CPVRManager::Deinit(), not only to CPVRManager::OnWake()

Finally, the IsStarted check should be removed from CPVRManager::TriggerSearchMissingChannelIcons().

**Review by ksooo — CHANGES_REQUESTED** [review]

File: xbmc/pvr/PVRManager.cpp (hunk 3 above)
Comment (xhaggi) [inline xbmc/pvr/PVRManager.cpp]: IMO every method should be self-contained. They need to validate the state if necessary. We should drop that check in OnWake.

File: xbmc/pvr/PVRManager.cpp (hunk 1 above)
Comment (xhaggi) [inline xbmc/pvr/PVRManager.cpp]: Sure will change the order.

File: xbmc/pvr/PVRManager.h (hunk 2 above)
Comment (xhaggi) [inline xbmc/pvr/PVRManager.h]: Okay

File: xbmc/pvr/PVRManager.cpp (hunk 3 above)
Comment (ksooo) [inline xbmc/pvr/PVRManager.cpp]: ok

**Comment by xhaggi** [comment]
@FernetMenta dropped the commit which removes the switch channel job

@ksooo mind taking another look at it. The last commit drops the globals in PVR manager and use the CServiceBroker instead.

**Review by ksooo — COMMENTED** [review]
No strong opinion on that, but what's wrong with g_* defines? The only "change" I see here is, that I now have to type more. For me, this is not an improvement, but a drawback.

**Comment by xhaggi** [comment]
We want to get rid of those globals and it's done now. Later we could think about to detach recordings, clients, timers etc from the PVRManager and provide them by the ServiceManager/ServiceBroker too.

**Comment by xhaggi** [comment]
jenkins build this please

**Comment by ksooo** [comment]
Sorry, I don't see the point here. But no -1 from my side. Just asking myself what the benefit/improvement/cleanup is.

**Comment by ksooo** [comment]
> Later we could think about to detach recordings, clients, timers etc from the PVRManager and provide them by the ServiceManager/ServiceBroker too.

 Haha. This could be done even more simple if we still had the defines.

old: #define g_PVRClients CServiceBroker::GetPVRMananger().Clients()
new: #define g_PVRClients CServiceBroker::GetPVRClients()

A one liner to change everything!

**Comment by ksooo** [comment]
> We want to get rid of those globals and it's done now.

 My understanding of our goal to "get rid of the gloabls" is not to remove defines, but do architectural changes. Imo, a good approach to "get rid of the globals" is to pass the "owning component" (the PVR Manager for example) to the "owned components" (the PVR recordings for example), so that the owned component knows its owner and can access it, without calling a global method. @FernetMenta what's your take on this?

**Comment by xhaggi** [comment]
> Imo, a good approach to "get rid of the globals" is to pass the "owning component" (the PVR Manager for example) to the "owned components" (the PVR recordings for example), so that the owned component knows its owner and can access it, without calling a global method.

That was my first thought, but passing the owner to each of them is not a nice approach for a service class. IMO all of them are services. My second thought was "why not detaching them from the PVR manager".  But I don't have a strong option on how to proceed here.

**Comment by xhaggi** [comment]
BTW passing the owner to the owned services does not really solve your "we have to write more code" issue. Each access to Recordings, Timers etc outside will result in the same call `CServiceBroker::GetPVRManager().GetTimers()`.

**Comment by xhaggi** [comment]
any objections or ready to merge?

After that I'll create another PR that drops the channel switch job as @FernetMenta suggested.

**Comment by xhaggi** [comment]
Noticed that the PVRManager still expose a `GetInstance()` method which delegates the call to the `ServiceBroker`. I've added another commit to drop that and use the `ServiceBroker` directly

jenkins build this please

**Comment by ksooo** [comment]
> Noticed that the PVRManager still expose a GetInstance() method which delegates the call to the ServiceBroker. I've added another commit to drop that and use the ServiceBroker directly

 Nice.

**Review by ksooo — APPROVED** [review]

**Comment by xhaggi** [comment]
jenkins build this please

**Comment by xhaggi** [comment]
WIN32 build errors are unrelated.

**Pull request merged**